### PR TITLE
htaccess: prevent accessing resources not managed by SCM

### PR DESCRIPTION
**Pull request by virtualtam**
See:
- https://en.internetwache.org/dont-publicly-expose-git-or-how-we-downloaded-your-websites-sourcecode-an-analysis-of-alexas-1m-28-07-2015/
- https://stackoverflow.com/questions/2530372/how-do-i-disable-directory-browsing
- https://httpd.apache.org/docs/current/mod/mod_rewrite.html

### CODE REVIEW

**Comment by nodiscc** [comment]
Personally I use apache `deny` directives on the whole VirtualHost for version control directories: https://github.com/nodiscc/srv01/blob/master/roles/apache2/files/etc_apache2_conf-available_prevent-versioncontrol-dir-access.conf.

I think a `deny` rule is more appropriate than a rewriterule for this kind of access attempt (it is not part of normal browsing behavior and will be logged as 401 Unauthorized, as it should be).

 - Should we use `deny` here instead?
   - Does it apply to apache only? nginx compatibility? (yeah I still have to switch)
 - Does the interaction of virtualhost-wide and `.htaccess` directives work as expected?
 - Should we document it anyway in https://github.com/shaarli/Shaarli/blob/master/doc/md/Server-configuration.md ?

**Comment by virtualtam** [comment]
> Should we document it anyway [...]?

Disabling directory listing is part of Apache best practices for a production / public-facing server. It is indeed worth mentioning in the documentation. Disabling access to Git metadata folders should also be done server-wide.

Think of this PR as a safety belt for lousy configured Apache HTTPD servers :)

Additionally, the two directives added in this PR (`Options`, `RewriteRule`) imply that `.htaccess` files are interpreted, and URL rewriting is enabled. These are already mentioned as prerequisites for running Shaarli.

> nginx compatibility

Nginx should not be impacted as:
- Docker images use release tarballs
- the Docker image configuration as well as the provided server configuration example prevent access to dotfiles / dotted folders and any PHP script other that `index.php`

> `401 Unauthorized`, as it should be

`401` implies that given the proper access rights / credentials, the resource could be accessed. This is not the case here as these files are not supposed to be served.

> Should we use `deny` here instead? 

AFAIK `deny` is only available to use when writing rules addressing the folder containing the `.htaccess` file; this is what we already use to prevent access to cache and data directories.

To address this issue with `deny`, one would have to embed it within a `DirectoryMatch` section, but this can only be done at the whole server or virtualhost level, hence the use of a rewrite rule.

**Comment by ArthurHoaro** [comment]
> Does it apply to apache only? nginx compatibility?

`.htaccess` are specific to Apache and there is no equivalent for nginx.

> Does the interaction of virtualhost-wide and .htaccess directives work as expected?

I don't see any issue here. `.htaccess` directives override the virtual host configuration, which makes sense here. If you want to configure everything at the VHost level, let's say for performances, you can add `AllowOverride None`.

File: .htaccess [hunk 1]
```diff
@@ -1,4 +1,12 @@
+# Disable directory listing
+Options -Indexes
+
 RewriteEngine On
+
+# Prevent accessing subdirectories not managed by SCM
+RewriteRule ^(.git|doxygen|vendor) - [F]
```
Comment (ArthurHoaro) [inline .htaccess]: `sandbox` folder could be added.

Comment (virtualtam) [reply .htaccess]: agreed, though I'm not sure one is supposed to run unit tests (nor install test dependencies) on a public-facing server :p

Comment (ArthurHoaro) [reply .htaccess]: I agree, I was just checking which untracked folders I had in my dev environment.

**Pull request merged**